packages/fuel-indexer-macros/src/helpers.rs

```diff
@@ -1,10 +1,107 @@
+use std::collections::HashSet;
+
 use crate::constant::*;
 use fuel_abi_types::program_abi::{ProgramABI, TypeDeclaration};
 use fuel_indexer_schema::parser::ParsedGraphQLSchema;
 use fuels_code_gen::utils::Source;
+use lazy_static::lazy_static;
 use quote::{format_ident, quote};
 use syn::Ident;
 
+lazy_static! {
+    /// Set of internal indexer entities.
+    pub static ref INTERNAL_INDEXER_ENTITIES: HashSet<&'static str> = HashSet::from([
+        "IndexMetadataEntity",
+    ]);
+
+    /// Set of types that implement `AsRef<[u8]>`.
+    pub static ref ASREF_BYTE_TYPES: HashSet<&'static str> = HashSet::from([
+        "Address",
+        "AssetId",
+        "Blob",
+        "BlockId",
+        "Boolean",
+        "Bytes",
+        "Bytes20",
+        "Bytes32",
+        "Bytes4",
+        "Bytes64",
+        "Bytes8",
+        "Charfield",
+        "ContractId",
+        "HexString",
+        "Json",
+        "MessageId",
+        "NoRelation",
+        "Nonce",
+        "Option<Address>",
+        "Option<AssetId>",
+        "Option<Blob>",
+        "Option<BlockId>",
+        "Option<Boolean>",
+        "Option<Bytes20>",
+        "Option<Bytes32>",
+        "Option<Bytes4>",
+        "Option<Bytes64>",
+        "Option<Bytes8>",
+        "Option<Bytes>",
+        "Option<Charfield>",
+        "Option<ContractId>",
+        "Option<HexString>",
+        "Option<Json>",
+        "Option<MessageId>",
+        "Option<NoRelation>",
+        "Option<Nonce>",
+        "Option<Salt>",
+        "Option<Signature>",
+        "Option<TxId>",
+        "Salt",
+        "Signature",
+        "TxId",
+    ]);
+
+    /// Set of external types that do not implement `AsRef<[u8]>`.
+    pub static ref EXTERNAL_FIELD_TYPES: HashSet<&'static str> = HashSet::from([
+        "Identity",
+        "Option<Identity>",
+        "Option<Tai64Timestamp>",
+        "Tai64Timestamp",
+    ]);
+
+    /// Set of field types that are currently unable to be used as a digest for SHA-256 hashing.
+    pub static ref NONDIGESTIBLE_FIELD_TYPES: HashSet<&'static str> = HashSet::from([
+        "Boolean",
+        "Identity"
+    ]);
+}
+
+/// Provides a TokenStream to be used as a conversion to bytes
+/// for external types; this is done because traits cannot be
+/// implemented on external types due to the orphan rule.
+pub fn to_bytes_method_for_external_type(
+    field_type_name: String,
+) -> proc_macro2::TokenStream {
+    match field_type_name.as_str() {
+        "Identity" => quote! { .0 },
+        "Tai64Timestamp" => quote! { .0.to_le_bytes() },
+        _ => panic!(
```

Comment:
- Maybe?
```suggestion
        _ => panic!("From<{field_type_name}> not implemented for AsRef<u8>.");
```
- Suggesting that the rust-y explanation might be easier to understand